src/StringTools/InternableString.cs

```diff
@@ -304,28 +304,60 @@ public override string ToString()
         /// <returns>A stable hashcode of the string represented by this instance.</returns>
         public override unsafe int GetHashCode()
         {
-            int hashCode = 5381;
+            uint hash = (5381 << 16) + 5381;
+            bool hashedOddNumberOfCharacters = false;
+
             fixed (char* charPtr = _inlineSpan)
             {
-                for (int i = 0; i < _inlineSpan.Length; i++)
-                {
-                    hashCode = unchecked(hashCode * 33 ^ charPtr[i]);
-                }
+                GetHashCodeHelper(charPtr, _inlineSpan.Length, ref hash, ref hashedOddNumberOfCharacters);
             }
             if (_spans != null)
             {
                 foreach (ReadOnlyMemory<char> span in _spans)
                 {
                     fixed (char* charPtr = span.Span)
                     {
-                        for (int i = 0; i < span.Length; i++)
-                        {
-                            hashCode = unchecked(hashCode * 33 ^ charPtr[i]);
-                        }
+                        GetHashCodeHelper(charPtr, span.Length, ref hash, ref hashedOddNumberOfCharacters);
                     }
                 }
             }
-            return hashCode;
+            return (int)(hash);
+        }
+
+        /// <summary>
+        /// Hashes a memory block specified by a pointer and length.
+        /// </summary>
+        /// <param name="charPtr">Pointer to the first character.</param>
+        /// <param name="length">Number of characters at <paramref name="charPtr"/>.</param>
+        /// <param name="hash">The running hash code.</param>
+        /// <param name="hashedOddNumberOfCharacters">True if the incoming <paramref name="hash"/> was calculated from an odd number of characters.</param>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static unsafe void GetHashCodeHelper(char* charPtr, int length, ref uint hash, ref bool hashedOddNumberOfCharacters)
+        {
+            if (hashedOddNumberOfCharacters && length > 0)
+            {
+                // If the number of characters hashed so far is odd, the first character of the current block completes
+                // the calculation done with the last character of the previous block.
+                hash ^= BitConverter.IsLittleEndian ? ((uint)*charPtr << 16) : *charPtr;
+                length--;
+                charPtr++;
+                hashedOddNumberOfCharacters = false;
+            }
+
+            // The loop hashes two characters at a time.
+            uint* ptr = (uint*)charPtr;
+            while (length >= 2)
+            {
+                length -= 2;
+                hash = ((hash << 5) + hash) ^ *ptr;
```

Comment:
How is this hashing two at a time? It looks like it's modifying the hash as I'd expect then xor'ing it with one character. Length goes down by two but index only increases by one? Would that miss the second half of the span?

Reply:
`ptr` is a `uint*` which is [syntactic sugar for an unsigned 32-bit int](https://docs.microsoft.com/dotnet/csharp/language-reference/builtin-types/integral-numeric-types#:~:text=%20Integral%20numeric%20types%20%28C%23%20reference%29%20%201,C%205%20%20language%20specification.%20%20More%20). `charPtr` though is a `char*` which is [syntactic sugar for a UTF-16 character](https://docs.microsoft.com/dotnet/csharp/language-reference/builtin-types/char). So `length` is a count of 16-bit things and "index"/`ptr` is a list of 32-bit things (so half the size if the list is even).